File: modulo04-agentes-autonomos/agent-evals-java/src/main/java/com/iadeva/evals/dataset/EvalDataset.java
package com.iadeva.evals.dataset;

// Equivalente ao eval_dataset.py — carrega e gerencia o dataset de avaliação

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.iadeva.evals.model.EvalScenario;
import jakarta.annotation.PostConstruct;
import org.springframework.core.io.ClassPathResource;
import org.springframework.stereotype.Component;

import java.io.IOException;
import java.util.List;

/**
 * Componente responsável por carregar os cenários de avaliação do arquivo JSON.
 * O dataset fica em src/main/resources/eval-dataset.json e é lido uma única vez na inicialização.
 */
@Component
public class EvalDataset {

    private final ObjectMapper objectMapper;
    private List<EvalScenario> scenarios;

    public EvalDataset(ObjectMapper objectMapper) {
        this.objectMapper = objectMapper;
    }

    /**
     * Carrega o dataset na inicialização do contexto Spring.
     * Falha imediatamente se o arquivo não existir ou estiver malformado.
     */
    @PostConstruct
    public void load() {
        try {
            var resource = new ClassPathResource("eval-dataset.json");
            scenarios = objectMapper.readValue(
                    resource.getInputStream(),
                    new TypeReference<List<EvalScenario>>() {}
            );
        } catch (IOException e) {
            throw new IllegalStateException("Falha ao carregar eval-dataset.json: " + e.getMessage(), e);
        }
    }

    /**
     * Retorna todos os cenários do dataset.
     *
     * @return lista imutável de cenários
     */
    public List<EvalScenario> getScenarios() {
        return List.copyOf(scenarios);
    }
}

File: modulo04-agentes-autonomos/agent-evals-go/dataset/eval_dataset.go
// Package dataset carrega o dataset de avaliação — equivalente a
// EvalDataset.java. O JSON é embarcado no binário via go:embed (equivalente
// a ClassPathResource, que empacota o arquivo dentro do jar), garantindo que
// esteja sempre disponível independente do diretório de trabalho.
package dataset

import (
	_ "embed"
	"encoding/json"
	"fmt"

	"agent-evals/model"
)

//go:embed eval-dataset.json
var evalDatasetJSON []byte

// EvalDataset é o componente responsável por expor os cenários de avaliação
// carregados na inicialização.
type EvalDataset struct {
	scenarios []model.EvalScenario
}

// Load carrega o dataset embarcado. Falha imediatamente (panic) se o JSON
// estiver malformado — equivalente ao IllegalStateException lançado por
// EvalDataset.load() em caso de falha.
func Load() *EvalDataset {
	var scenarios []model.EvalScenario
	if err := json.Unmarshal(evalDatasetJSON, &scenarios); err != nil {
		panic(fmt.Sprintf("falha ao carregar eval-dataset.json: %v", err))
	}
	return &EvalDataset{scenarios: scenarios}
}

// Scenarios retorna todos os cenários do dataset.
func (d *EvalDataset) Scenarios() []model.EvalScenario {
	result := make([]model.EvalScenario, len(d.scenarios))
	copy(result, d.scenarios)
	return result
}
